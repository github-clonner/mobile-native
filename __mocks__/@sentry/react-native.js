export const captureMessage = jest.fn();
export default {};
