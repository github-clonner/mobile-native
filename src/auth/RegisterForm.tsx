//@ts-nocheck
import React, { Component } from 'react';

import {
  Text,
  View,
  ScrollView,
  Linking,
  Alert,
  TouchableOpacity,
  SafeAreaView,
  LayoutAnimation,
} from 'react-native';

import authService from '../auth/AuthService';
import { observer, inject } from 'mobx-react';
import { CheckBox } from 'react-native-elements';

import i18n from '../common/services/i18n.service';
import sessionService from '../common/services/session.service';
import delay from '../common/helpers/delay';
import apiService from '../common/services/api.service';
import Input from '../common/components/Input';
import Icon from 'react-native-vector-icons/MaterialIcons';
import Button from '../common/components/Button';
import { DISABLE_PASSWORD_INPUTS } from '../config/Config';
import ThemedStyles from '../styles/ThemedStyles';
import PasswordValidator from '../common/components/PasswordValidator';
import validatePassword from '../common/helpers/validatePassword';

import type { registerParams } from '../auth/AuthService';
import logService from '../common/services/log.service';

/**
 * Register Form
 */
@inject('user')
@observer
class RegisterForm extends Component {
  state = {
    error: {},
    password: '',
    passwordFocused: false,
    username: '',
    confirmPassword: '',
    email: '',
    termsAccepted: false,
    exclusive_promotions: false,
    inProgress: false,
  };

  validatePassword(value) {
    let error = this.state.error;
    if (this.state.password !== value) {
      error.confirmPasswordError = i18n.t('auth.confirmPasswordError');
    } else {
      error.confirmPasswordError = '';
    }

    error.invalidPasswordError = !validatePassword(value).all
      ? i18n.t('auth.invalidPassword')
      : '';

    this.setState({ error });
  }

  validateTerms(value) {
    let error = this.state.error;
    if (!this.state.termsAccepted && this.state.username.length > 4) {
      error.termsAcceptedError = i18n.t('auth.termsAcceptedError');
    } else {
      error.termsAcceptedError = '';
    }
    this.setState({ termsAccepted: !!value, error });
  }

  setUsername = (username) => this.setState({ username });
  setEmail = (email) => this.setState({ email });
  setPassword = (password) => this.setState({ password });
  setConfirmPassword = (confirmPassword) => this.setState({ confirmPassword });

  getFormBody = () => {
    const CS = ThemedStyles.style;
    return (
      <ScrollView
        style={[CS.flexContainer, CS.marginTop2x]}
        contentContainerStyle={[CS.paddingHorizontal4x, CS.paddingBottom5x]}>
        <SafeAreaView style={CS.flexContainer}>
          <TouchableOpacity
            onPress={this.props.onBack}
            style={CS.marginBottom3x}>
            <Icon
              size={34}
              name="keyboard-arrow-left"
              style={CS.colorSecondaryText}
            />
          </TouchableOpacity>
          <Text
            style={[
              CS.marginBottom3x,
              CS.textCenter,
              CS.titleText,
              CS.colorPrimaryText,
            ]}>
            {i18n.t('auth.join')}
          </Text>
          <Text style={[CS.colorAlert, CS.textCenter]}>
            {this.state.error.termsAcceptedError}
          </Text>
          {this.state.passwordFocused ? (
            <PasswordValidator password={this.state.password} />
          ) : (
            <>
              <Input
                style={CS.marginBottom2x}
                placeholder={i18n.t('auth.username')}
                onChangeText={this.setUsername}
                value={this.state.username}
                editable={!this.state.inProgress}
                testID="registerUsernameInput"
              />
              <Input
                style={CS.marginBottom2x}
                placeholder={i18n.t('auth.email')}
                onChangeText={this.setEmail}
                value={this.state.email}
                editable={!this.state.inProgress}
                testID="registerEmailInput"
              />
            </>
          )}

          <Input
            style={CS.marginBottom2x}
            placeholder={i18n.t('auth.password')}
            secureTextEntry={!DISABLE_PASSWORD_INPUTS} // e2e workaround
            onChangeText={this.setPassword}
            value={this.state.password}
            editable={!this.state.inProgress}
            onFocus={this.focusPassword}
            onBlur={this.blurPassword}
            testID="registerPasswordInput"
          />
          <Input
            placeholder={i18n.t('auth.confirmpassword')}
            secureTextEntry={!DISABLE_PASSWORD_INPUTS} // e2e workaround
            onChangeText={this.setConfirmPassword}
            value={this.state.confirmPassword}
            editable={!this.state.inProgress && this.state.password}
            testID="registerPasswordConfirmInput"
          />
          <CheckBox
            containerStyle={CS.checkbox}
            title={
              <Text style={[CS.colorSecondaryText, CS.fontL]}>
                {i18n.t('auth.accept')}{' '}
                <Text
                  style={CS.link}
                  onPress={() =>
                    Linking.openURL('https://www.minds.com/p/terms')
                  }>
                  {i18n.t('auth.termsAndConditions')}
                </Text>
              </Text>
            }
            checked={this.state.termsAccepted}
            onPress={this.check}
            disabled={this.state.inProgress}
            testID="checkbox"
          />

          <View style={(CS.flexContainer, CS.paddingTop2x)}>
            <Button
              onPress={() => this.onPressRegister()}
              borderRadius={2}
              containerStyle={[CS.button, CS.fullWidth]}
              textStyle={CS.buttonText}
              loading={this.state.inProgress}
              loadingRight={true}
              disabled={this.state.inProgress}
              text={i18n.t('auth.createChannel')}
              testID="registerCreateButton"
            />
            <Text
              style={[
                CS.subTitleText,
                CS.colorSecondaryText,
                CS.centered,
                CS.marginTop2x,
              ]}>
              {i18n.to('auth.alreadyHaveAccount', null, {
                login: (
                  <Text style={[CS.link, CS.fontL]} onPress={this.props.onBack}>
                    {i18n.t('auth.login')}
                  </Text>
                ),
              })}
            </Text>
          </View>
        </SafeAreaView>
      </ScrollView>
    );
  };

  check = () => {
    this.setState({ termsAccepted: !this.state.termsAccepted });
  };

  render() {
    const CS = ThemedStyles.style;
    return (
      <View style={[CS.flexContainerCenter, CS.backgroundPrimary]}>
        {this.getFormBody()}
      </View>
    );
  }

  focusPassword = () => {
    this.setState({ passwordFocused: true });
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
  };

  blurPassword = () => {
    this.setState({ passwordFocused: false });
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
  };

  /**
   * On press register
   */
  async onPressRegister() {
    this.validatePassword(this.state.confirmPassword);

    if (!this.state.termsAccepted) {
      return Alert.alert(i18n.t('ops'), i18n.t('auth.termsAcceptedError'));
    }

    if (this.state.error.confirmPasswordError) {
      return Alert.alert(i18n.t('ops'), this.state.error.confirmPasswordError);
    }

    if (this.state.error.invalidPasswordError) {
      return Alert.alert(i18n.t('ops'), this.state.error.invalidPasswordError);
    }

    this.setState({ inProgress: true });

    try {
      const params = {
        username: this.state.username,
        email: this.state.email,
        password: this.state.password,
        exclusive_promotions: this.state.exclusive_promotions,
      } as registerParams;
      await authService.register(params);
      sessionService.setInitialScreen('OnboardingScreen');
      await apiService.clearCookies();
      await delay(100);
      try {
        await authService.login(this.state.username, this.state.password);
      } catch (err) {
        try {
          await authService.login(this.state.username, this.state.password);
        } catch (error) {
          Alert.alert(i18n.t('ops'), i18n.t('auth.failedToLoginNewAccount'));
          logService.exception(error);
        }
      }
    } catch (err) {
      Alert.alert(i18n.t('ops'), err.message);
      logService.exception(err);
    }

    this.setState({ inProgress: false });
  }
}

export default RegisterForm;
