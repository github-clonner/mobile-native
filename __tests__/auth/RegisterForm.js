import 'react-native';
import React from 'react';
import { Alert } from "react-native";
import { shallow } from 'enzyme';
import { Button, CheckBox } from 'react-native-elements';

import RegisterForm from '../../src/auth/RegisterForm';
import authService from '../../src/auth/AuthService';

jest.mock('../../src/auth/AuthService');
jest.mock('../../src/auth/UserStore');


Alert.alert = jest.fn();

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import UserStore from '../../src/auth/UserStore';

describe('RegisterForm component', () => {
  beforeEach(() => {
    authService.register.mockClear();
    Alert.alert.mockClear();
  });

  it('should renders correctly', () => {
    const userStore = new UserStore();
    const registerForm = renderer.create(
      <RegisterForm user={userStore} />
    ).toJSON();
    expect(registerForm).toMatchSnapshot();
  });

  it('should call auth services register', async () => {

    authService.register.mockResolvedValue();

    const userStore = new UserStore();

    const wrapper = shallow(
      <RegisterForm user={userStore} />
    );

    const render = wrapper.dive();

    // find the text inputs
    let inputs = render.find('Input');

    // should have 3 inputs
    expect(inputs.length).toBe(4);

    // simulate user input
    inputs.at(0).simulate('changeText', 'myFancyUsername');
    inputs.at(1).simulate('changeText', 'my@mail.com');
    inputs.at(3).simulate('changeText', '$0453Article');
    inputs.at(2).simulate('changeText', '$0453Article');

    // update component (password confirmation is shown after the password field is set)
    await wrapper.update();

    // simulate user input for paddword confirmation

    // simulate press checkbox
    await render.find(CheckBox).at(0).simulate('press');

    // simulate press register
    await render.find('Button').at(0).simulate('press');

    // expect auth service register to be called once
    expect(authService.register).toBeCalled();

  });

  it('should warn the user if the password confirmation is different', async () => {
    const userStore = new UserStore();

    const wrapper = shallow(
      <RegisterForm user={userStore} />
    );

    const render = wrapper.dive();

    // find the text inputs
    let inputs = render.find('Input');

    // should have 4 inputs
    expect(inputs.length).toBe(4);

    // simulate user input
    inputs.at(0).simulate('changeText', 'myFancyUsername');
    inputs.at(1).simulate('changeText', 'my@mail.com');
    inputs.at(2).simulate('changeText', 'somepassword');
    // update component (password confirmation is shown after the password field is set)
    await wrapper.update();

    // update the inputs search
    inputs = render.find('Input');

    // simulate user input for paddword confirmation
    inputs.at(3).simulate('changeText', 'ohNoItIsDifferent');

    // simulate press register
    await render.find('Button').at(0).simulate('press');

    // should call alert
    expect(Alert.alert).toBeCalled();

    // with error message
    expect(Alert.alert.mock.calls[0][1]).toEqual('You should accept the terms and conditions');
  });

  it('should warn the user if the terms and conditions are not accepted', async () => {
    const userStore = new UserStore();

    const wrapper = shallow(
      <RegisterForm user={userStore}/>
    );

    const render = wrapper.dive();

    // find the text inputs
    let inputs = render.find('Input');

    // should have 4 inputs
    expect(inputs.length).toBe(4);

    // simulate user input
    inputs.at(0).simulate('changeText', 'myFancyUsername');
    inputs.at(1).simulate('changeText', 'my@mail.com');
    inputs.at(2).simulate('changeText', 'somepassword');
    // update component (password confirmation is shown after the password field is set)
    await wrapper.update();

    // update the inputs search
    inputs = render.find('Input');

    // simulate user input for paddword confirmation
    inputs.at(3).simulate('changeText', 'somepassword');

    // simulate press register
    await render.find('Button').at(0).simulate('press');

    // should call alert
    expect(Alert.alert).toBeCalled();

    // with error message
    expect(Alert.alert.mock.calls[0][1]).toEqual('You should accept the terms and conditions');
  });
});
