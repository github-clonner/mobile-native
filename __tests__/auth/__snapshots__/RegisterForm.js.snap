// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm component should renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Array [
        undefined,
        undefined,
      ]
    }
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <RCTSafeAreaView
        emulateUnlessSupported={true}
      >
        <View
          accessible={true}
          focusable={false}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 34,
                },
                undefined,
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
        <Text
          style={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          Join the Revolution
        </Text>
        <Text
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        />
        <View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#AEB0B8",
                      "fontFamily": "Roboto",
                      "fontSize": 16,
                      "fontWeight": "600",
                      "marginBottom": 5,
                    },
                    undefined,
                  ]
                }
              >
                Username
              </Text>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            editable={false}
            onChangeText={[Function]}
            placeholder=""
            placeholderTextColor="#444"
            rejectResponderTermination={true}
            returnKeyType="done"
            style={
              Array [
                undefined,
                undefined,
              ]
            }
            testID="registerUsernameInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#AEB0B8",
                      "fontFamily": "Roboto",
                      "fontSize": 16,
                      "fontWeight": "600",
                      "marginBottom": 5,
                    },
                    undefined,
                  ]
                }
              >
                Email
              </Text>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            editable={false}
            onChangeText={[Function]}
            placeholder=""
            placeholderTextColor="#444"
            rejectResponderTermination={true}
            returnKeyType="done"
            style={
              Array [
                undefined,
                undefined,
              ]
            }
            testID="registerEmailInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#AEB0B8",
                      "fontFamily": "Roboto",
                      "fontSize": 16,
                      "fontWeight": "600",
                      "marginBottom": 5,
                    },
                    undefined,
                  ]
                }
              >
                Password
              </Text>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            editable={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder=""
            placeholderTextColor="#444"
            rejectResponderTermination={true}
            returnKeyType="done"
            secureTextEntry={true}
            style={
              Array [
                undefined,
                undefined,
              ]
            }
            testID="registerPasswordInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#AEB0B8",
                      "fontFamily": "Roboto",
                      "fontSize": 16,
                      "fontWeight": "600",
                      "marginBottom": 5,
                    },
                    undefined,
                  ]
                }
              >
                Confirm Password
              </Text>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            editable={true}
            onChangeText={[Function]}
            placeholder=""
            placeholderTextColor="#444"
            rejectResponderTermination={true}
            returnKeyType="done"
            secureTextEntry={true}
            style={
              Array [
                undefined,
                undefined,
              ]
            }
            testID="registerPasswordConfirmInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#fafafa",
              "borderColor": "#ededed",
              "borderRadius": 3,
              "borderWidth": 1,
              "margin": 5,
              "marginLeft": 10,
              "marginRight": 10,
              "opacity": 1,
              "padding": 10,
            }
          }
          testID="checkbox"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#bfbfbf",
                    "fontSize": 24,
                  },
                  Object {
                    "minWidth": 24,
                  },
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
            <Text
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            >
              I have read and accept the
               
              <Text
                onPress={[Function]}
              >
                terms of use
              </Text>
            </Text>
          </View>
        </View>
        <View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#5DBAC0",
                "borderRadius": 36,
                "opacity": 1,
              }
            }
            testID="registerCreateButton"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "white",
                  },
                  undefined,
                ]
              }
            >
               
              Join Minds Now
               
            </Text>
          </View>
          <Text
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
          >
            Already have an account? 
            <Text
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            >
              Login
            </Text>
            
          </Text>
        </View>
      </RCTSafeAreaView>
    </View>
  </RCTScrollView>
</View>
`;
